File: tests/record/set_explicit/record_set_explicit.ml
open Optitrust
open Target

(* let _ = Flags.check_validity := true *)

let _ = Run.script_cpp (fun _ ->
  (* Example with detach of initialization *)
  !! Record.set_explicit [cVarDef "p"];
  (* Another example with more complex initializers *)
  !! Record.set_explicit [sInstr "obj a = "];
  !! Record.set_explicit [sInstr "a.speed = "];
  (* Another example with a more complex right-hand side *)
  !! Record.set_explicit [cVarDef "u"];
  (* Example without detach *)
  !! Record.set_explicit [sInstr "b = p"];
  (* Example with const *)
  !! Record.set_explicit [sInstr "*v = c"];
)

File: tests/record/set_explicit/record_set_explicit.cpp

typedef struct {
    int x;
    int y; }
  vect;

typedef struct {
  int weight;
  vect pos;
  vect speed;
} obj;

vect f() {
  return {1,1};
}

int main() {
  vect p = {0,0};
  vect b;
  b = p;

  obj a = {0,{0,0},{0,0}};
  vect u = a.pos;

  const obj c = {0,{0,0},{0,0}};
  obj *const v = &a;
  *v = c;
}
